gflownet/utils/batch.py

```diff
@@ -668,22 +686,51 @@ def get_logprobs(
             valid. The flag is False if the corresponding logprob value is a zero / None
             placefolder.
         """
+        if not self._logprobs_available:
+            self._compute_logprobs()
 
         if backward:
-            if self.logprobs_backward[0] is None:
-                return None, None
-            logprobs = tfloat(
-                self.logprobs_backward, float_type=self.float, device=self.device
-            )
-            valids = tbool(self.logprobs_backward_valid, device=self.device)
+            return self.logprobs_backward, self.logprobs_backward_valid
         else:
-            if self.logprobs_forward[0] is None:
-                return None, None
-            logprobs = tfloat(
-                self.logprobs_forward, float_type=self.float, device=self.device
+            return self.logprobs_forward, self.logprobs_forward_valid
+
+    def _compute_logprobs(self):
+        """
+        Convert all logprobs and their validity flags to tensors and update validity flags
+        """
+        self.logprobs_forward, self.logprobs_forward_valid = (
+            self._logprobs_to_clean_tensor(
+                self.logprobs_forward, self.logprobs_forward_valid
             )
-            valids = tbool(self.logprobs_forward_valid, device=self.device)
-        return logprobs, valids
+        )
+        self.logprobs_backward, self.logprobs_backward_valid = (
+            self._logprobs_to_clean_tensor(
+                self.logprobs_backward, self.logprobs_backward_valid
+            )
+        )
+        self._logprobs_available = True
+
+    def _logprobs_to_clean_tensor(self, logprobs, logprobs_valid):
+        """
+        Convert logprobs and logprobs_valid to tensors and update validity flags
+        """
+        if any(x is None for x in logprobs):
+            logprobs_clean = []
+            for idx, lp in enumerate(logprobs):
+                if lp is None:
+                    logprobs_valid[idx] = False
+                    logprobs_clean.append(
+                        tfloat(0.0, device=self.device, float_type=self.float)
+                    )
+                else:
+                    logprobs_clean.append(lp)
+        else:
+            logprobs_clean = logprobs
```

Comment:
I think there are a bunch of inefficient things here:
- Two for loops that can be avoided
- Tensor creation scalar by scalar
I believe all this can be improved through indexing with the help of numpy, which allows creating an array even though it's got None's in it. I already used this trick [here](https://github.com/alexhernandezgarcia/gflownet/blob/main/gflownet/gflownet.py#L407).
I'm making a suggestion directly here but I'm not testing it so even the syntax may be wrong. It's just for illustration, but please check it yourself.  
```suggestion
        logprobs_clean = np.array(logprobs)
        logprobs_clean[logprobs_clean == None] = 0.0
```

Reply:
The problem is that in the general case, some of the elements in the `logprobs ` list are tensors with attached gradients to them and some others are Nones. So, I cannot convert the whole `logprobs ` into a numpy array as I will lose gradients. I also cannot convert it into a tensor because torch throws an error because of the Nones (you cannot do `torch.tensor([tensor, None, tensor, None]`). Therefore, iterating over the list and converting one-by-one is the only possible solution here

Reply:
Oh I see! I had not thought of the gradients, but now I think about it, I'm surprised this works at all because the line below would destroy the gradients too!
```
logprobs_clean = tfloat(
    logprobs_clean, float_type=self.float, device=self.device
)
```

Reply:
With this in mind, then how about the following:
```
logprobs_clean = [logp if logp not None else 0.0 for logp in logprobs]
```

Reply:
> No, it will not :) It will call inside the torch.stack over all the one-element tensors inside the logprobs_clean and this operation preserves the connection to the computational graphs
Uhm ok, I had done a quick independent test and I could not see the gradients any more. But it was because I was not calling stack like you noted.

Then my other suggestion would work, right? Namely:
```
logprobs_clean = [logp if logp not None else 0.0 for logp in logprobs]
```

Reply:
Unfortunately, this won't work either, because some of the logp elements are one-element tensors, so tfloat will try to stack them with the floats (zeros which you put here) and throw this error: 
<img width="596" height="137" alt="Screenshot 2025-09-19 at 18 10 49" src="https://github.com/user-attachments/assets/eb57ca0f-db58-40a6-b63c-5cf9c1a62d9d" />

Reply:
Ok, my final attempt:
```
logprobs_clean = [logp if logp not None else torch.tensor(0.0) for logp in logprobs]
```
I guess using tfloat instead of just torch.tensor...

Reply:
That would work, but I also need to update `logprobs_valid` simultaneously as I'm creating `logprobs_clean`

Reply:
Ok let's merge it and simplify this later.